serde_codegen/src/de.rs

```diff
@@ -420,14 +421,16 @@ fn deserialize_seq(
                         })
                     }
                 };
-                quote_stmt!(cx,
+                let assign = quote_stmt!(cx,
                     let $name = match $visit {
                         Some(value) => { value },
                         None => {
-                            return Err(_serde::de::Error::end_of_stream());
+                            return Err(_serde::de::Error::invalid_length($index_in_seq));
```

Comment:
`$i` is the index in the struct, not index in the input sequence. The second unit test tests this. With `$i` it would return `InvalidLength(2)` instead of `InvalidLength(1)`.